internal/compiler-bridge/src/main/scala/xsbt/CompilerInterface.scala

```diff
@@ -115,7 +115,17 @@ private final class CachedCompiler0(args: Array[String], output: Output, initial
   def infoOnCachedCompiler(compilerId: String): String =
     s"[zinc] Running cached compiler $compilerId for Scala compiler $versionString"
 
-  def run(
+  // This is kept for compatibility purpose only.
+  override def run(
+      sources: Array[File],
+      changes: DependencyChanges,
+      callback: AnalysisCallback,
+      log: Logger,
+      delegate: Reporter,
+      progress: CompileProgress
+  ): Unit = ???
```

Comment:
How does this work for compatibility?  Should this be calling the other run somehow?

Reply:
Zinc 1.4.x should never use this method to enter into this (source-compiled) bridge. If we ever reach in here that means there's a bug in pattern match behavior, which I want to fail spectacularly.

Reply:
Could you capture that in the thrown exception, please?  If there's more to say then appropriate for the message, then as a comment.  Thanks!

Reply:
Why does `CachedCompiler2` extends `CachedCompiler1` then?

Reply:
Sure. I'll make those changes.

Reply:
Ideally I would split `CachedCompiler2` and `CachedCompiler` but I think it's kind of annoying for that one because compiler cache is being used by sbt:

```scala
      compilerCache := state.value get Keys.stateCompilerCache getOrElse CompilerCache.fresh,
```

and the implementation is tied to `CachedCompiler` in https://github.com/sbt/zinc/blob/v1.4.0-M6/internal/zinc-compile-core/src/main/java/xsbti/compile/CompilerCache.java  + https://github.com/sbt/zinc/blob/v1.4.0-M6/internal/zinc-compile-core/src/main/scala/sbt/internal/inc/CompilerCache.scala.

So in `CachedCompiler2` would have two `run` method. One that works, and the other that would throw an error.

Reply:
In https://github.com/sbt/zinc/pull/829/commits/4774e53fb157b9aee24dce605897ebf417d543a7 I moved the error throwing to `CachedCompiler2#run`'s default method.